File: leetcode/product-of-array-except-self.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<int> productExceptSelf(vector<int>& nums)
    {
        int zeroCount = 0;
        int multiple = 1;

        vector<int> res;

        for (auto n : nums) {
            zeroCount += (n == 0);
            if (n != 0) {
                multiple *= n;
            }
        }

        if (zeroCount < 2) {
            for (auto n : nums) {
                if (zeroCount > 0 && n != 0) {
                    res.push_back(0);
                } else {
                    res.push_back(multiple / (n == 0 ? 1 : n));
                }
            }
        } else {
            res.resize(nums.size());
        }

        return res;
    }
};

File: leetcode/product-of-array-except-self.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length = len(nums)
        prod = 1

        no_of_zeros = 0

        for i in range(length):
            if nums[i] != 0:
                prod *= nums[i]
            else:
                no_of_zeros += 1

        res = [0] * length

        if no_of_zeros <= 1:
            for i in range(length):
                n = nums[i]
                if n == 0:
                    res[i] = prod
                elif no_of_zeros == 0:
                    res[i] = prod // n

        return res
